### PR TITLE
Measure how much time placeholders spend on screen

### PR DESCRIPTION
**Pull request by gravityrail**
TL/DR: monitor how much time Calypso users spend looking at spinners so that we can focus on performance.

----

With the advent of single page apps everything is "instant"... or is it?

Instead of one spinning indicator on page load we can have dozens, and while we avoid the trap of a frustrated user clicking the button over and over again because they're not sure if things are working, we're still not doing any better in raw performance (sometimes worse, due to the bottleneck of all the API calls we're doing to the same domain and the volume of data retrieved).

Furthermore, optimisation is hard because we don't currently know where users are waiting. Should we prefetch posts for the reader? Should we only fetch 5 posts instead of 10? Should we batch together certain API calls? How does that affect perceived load time?

In order to make this "waiting time" visible, this PR creates a library called `perfmon`, which records how much time one or more loading indicators spend on the screen (currently timings are only sent to Google Analytics, but the goal is to get them into statsd or Kibana pronto). This includes those flashing grey divs, stats boxes, and the "pulsing dot", but could be extended to any component.

A key thing to note about my approach is that I wanted to avoid directly coupling with any components - the last thing I wanted to add for an experiment is a helper class mixin over the place, and remembering to add it to any new placeholder-aware components in the future would be error prone.

Instead I monitor the DOM itself for anything that looks like a placeholder (being careful not to do anything that would noticeably hurt performance), figure out whether it's visible on the screen, and if so I can record how long it was there for.

Here's how it looks in Google Analytics, showing the cumulative time spent looking at placeholders in various sections of the product:

timing-sample

And here's how it looks in Graphite:

graphite-placeholders

How it works:

- A MutationObserver is created which monitors DOM childList and class attribute changes within `div#wpcom`.
- It matches classes on these altered nodes (and their children) against a very rudimentary list of strings (anything containing 'placeholder', 'is-loading' or 'pulsing-dot is-active'). Any nodes that match are recorded as "activePlaceholders".
- It checks to see if any activePlaceholders are in the viewport. The event timing is the duration between the first placeholder appearing in the viewport and the last one disappearing.
- Placeholders can also be scrolled into view - it monitors for 'scroll' events with useCapture = true (debounced to 200ms) so that we can re-check if any known placeholders have now appeared.
- If the user navigates to another URL, an event is fired for any active placeholders and a new countdown begins, so that wait time is correctly attributed to each URL.

I've done some performance testing and on Chrome on my Macbook Pro it adds about 10-20ms to first 15 seconds or so after the "reader" page loads (a very placeholder-heavy page), which works out to be < 1% performance degradation. 

You can see how many active placeholders (visible and non visible) there are by running this in your console:

```
localStorage.setItem('debug', 'calypso:perfmon')
```

This has been tested in the latest Chrome, Firefox and Safari on a Mac and Edge on Windows.

At this point I think this code is ready for review by others. It would be great to get this into Horizon and see what real world usage looks like.

Future work:
- ~~Hide behind a configuration variable!~~ DONE!
- ~~Log data in an internal API for filtering and aggregation - need to make a post somewhere to propose how such an API would look and how data should be stored.~~ DONE! (thanks @josephscott)
- Capture more information about which components are loading slowly (it's hard to go from DOM->React, even with a data-reactid)
- Figure out a better way to record the page URL than memoizing it when someone calls analytics.pageView.record :/
- ~~Perhaps distinguish between scroll-to events and those triggered by DOM changes~~
- Tests, ~~README~~, etc :)

cc @beaulebens @roccotripaldi @martinremy

### CODE REVIEW

File: client/lib/perfmon/index.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ */
+import each from 'lodash/collection/each';
+import remove from 'lodash/array/remove';
+import debounce from 'lodash/function/debounce';
+
+var debug = require( 'debug' )( 'calypso:perfmon' );
+
+const PLACEHOLDER_CLASSES = [
+	'placeholder',
+	'pulsing-dot is-active'
```
Comment (gravityrail) [inline client/lib/perfmon/index.js]: This is brittle since order of classes can change, kept it as a single string for simplicity but probably should be an array.

**Comment by roccotripaldi** [comment]
Impressive! I don't fully grok the code yet as this is the first I've heard of `MutationObserver`. But I like your instincts here and I think this approach could give good insight into slow queries.

I'll tinker some more with it as the week goes on, but it seems to be working! I've seen a few console errors pop up, but I haven't been able to consistently reproduce them, so i can't say for certain if they relate to Perfmon.

**Comment by gravityrail** [comment]
@roccotripaldi I added a few bugfixes for some console errors I was seeing. Could you update and tell me if there's still happening?

File: client/lib/perfmon/index.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ 	// and use these to trigger checks for visible placeholders (and, in the case of mutations,
+ 	// to record new placeholders and remove nodes that are no longer placeholders)
+	observeDomChanges: function( MutationObserver ) {
+
+		// if anything scrolls, check if any of our placeholder elements are in view,
+		// but not more than a few times a second
+		window.addEventListener('scroll', debounce(this.checkForVisiblePlaceholders.bind( this, 'scroll' ), 200), true);
+
+		// if the user navigates, stop the current event and proceed to the next one
+		page( function( context, next ) {
```
Comment (blowery) [inline client/lib/perfmon/index.js]: this depends on init() being called before all the routes are set up. Otherwise, it'll be added to the end of the chain and may never be called...

File: client/lib/perfmon/index.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+			// send "placeholder-wait-navigated" event
+			this.checkForVisiblePlaceholders( 'navigate' );
+			next();
+		}.bind( this ) );
+
+		// this is fired for matching mutations (childList and class attr changes)
+		var observer = new MutationObserver(function(mutations, observer) {
+			
+			// record all the nodes that match our placeholder classes in the "activePlaceholders" array
+			mutations.forEach( this.recordPlaceholders.bind( this ) );
```
Comment (blowery) [inline client/lib/perfmon/index.js]: `mutations.forEach( this.recordPlaceholders, this )` may be a touch faster and will generate less garbage for the GC to clean up

File: client/lib/perfmon/index.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	},
+
+	// check if there are any placeholders on the screen,
+	// and trigger a timing event when all the placeholders are gone or the user
+	// has navigated
+	checkForVisiblePlaceholders: function( trigger ) {
+
+		// determine how many placeholders are active in the viewport
+		var visibleCount = 0;
+		activePlaceholders.forEach( function( node ) {
```
Comment (blowery) [inline client/lib/perfmon/index.js]: reduce might be more idiomatic here

so `const visibleCount = activePlaceholders.reduce( function( count, node ) { return count + ( this.isElementVisibleInViewport( node ) ? 1 : 0 ); }.bind(this), 0 )`

File: client/lib/perfmon/index.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	'is-loading'
+];
+
+const PLACEHOLDER_MATCHER = PLACEHOLDER_CLASSES.map(function(clazz) { return `[class*='${clazz}']`; }).join(', ');
+const OBSERVE_ROOT = document.getElementById('wpcom');
+
+let activePlaceholders = [];
+let placeholdersVisibleStart = null;
+
+var perfmon = {
```
Comment (blowery) [inline client/lib/perfmon/index.js]: Why an object? You could use module level functions and avoid all the binding you end up having to do throughout this module.

File: client/lib/perfmon/index.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+	init: function() {
+		var MutationObserver = window.MutationObserver || window.WebKitMutationObserver;
```
Comment (blowery) [inline client/lib/perfmon/index.js]: Should you verify that init hasn't already been called?

File: client/lib/perfmon/index.js [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+	isPlaceholder: function( node ) {
+		var className = node.className;
+		return className && className.indexOf && PLACEHOLDER_CLASSES.some( function( clazz ) { 
+			return (className.indexOf( clazz ) >= 0); 
+		} );
+	},
+
+	// adapted from http://stackoverflow.com/questions/123999/how-to-tell-if-a-dom-element-is-visible-in-the-current-viewport/7557433#7557433
+	isElementVisibleInViewport: function( node ) {
```
Comment (blowery) [inline client/lib/perfmon/index.js]: this would be really nice as it's own thing in /client/lib/

File: client/lib/perfmon/index.js [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+		each( nodes, function( node ) {
+
+			if ( this.isPlaceholder( node ) ) {
+				this.recordPlaceholderNode( node );
+			}
+
+			// we need to find matching children because the mutation observer
+			// only fires for the top element of an added subtree
+			if ( node.querySelectorAll ) {
+				// funky syntax because NodeList walks like an array but doesn't quack like one
```
Comment (blowery) [inline client/lib/perfmon/index.js]: could just use `each` like you did up on line 149?

**Comment by blowery** [comment]
This is pretty neat! I'm a little concerned with the perf on lower end devices (older laptop running IE11 or a Nexus 5 maybe?), and I think you could clean up all those binds by using module level functions instead of hanging them all off the perfmon object. It doesn't appear that you're using it to store any state that couldn't be served by a module level variable, and it would save creating some garbage for the GC on every mutation.

File: client/lib/perfmon/index.js (hunk 3 above)
Comment (gravityrail) [inline client/lib/perfmon/index.js]: Fixed by moving functions out of hash.

File: client/lib/perfmon/index.js (hunk 4 above)
Comment (gravityrail) [inline client/lib/perfmon/index.js]: Great suggestion. Done :)

File: client/lib/perfmon/index.js (hunk 5 above)
Comment (gravityrail) [inline client/lib/perfmon/index.js]: Changed, yeah it felt weird to me too. Much cleaner now.

File: client/lib/perfmon/index.js (hunk 6 above)
Comment (gravityrail) [inline client/lib/perfmon/index.js]: Fixed in https://github.com/Automattic/wp-calypso/commit/72b498ca9b1cf8ab5466617c71de7ee5e50f0646

File: client/lib/perfmon/index.js (hunk 8 above)
Comment (gravityrail) [inline client/lib/perfmon/index.js]: Good suggestion. Fixed in https://github.com/Automattic/wp-calypso/commit/c17b1fcc529bbe615131736f0bac4fd7ddbe0d86

**Comment by gravityrail** [comment]
Thanks for all the great suggestions @blowery. In terms of performance, I just tried it on a Nexus 5 using Browserstack and it seems like it doesn't have a noticeable effect. That said, the Browserstack tab kept crashing when I was trying to use the Chrome inspector with the mobile device, so I didn't get to dig too deep. I also tested on IE11 and Calypso in general seems sluggish, but it didn't seem _more_ sluggish with this enabled. 

For example, the CSS flashing and existing infinite scroll behaviours use much more CPU because they're evaluating the position of more DOM elements (or so it seems from a cursory evaluation).

That said, if you see any performance corner cases (I'm sure they exist!) then I'd love to hear about them. I've mostly been focused on the reader, theme and stats pages since they seem to have the most placeholder elements.

**Comment by lancewillett** [comment]
:ship:  Things are looking good in testing. I see:

screen shot 2016-01-08 at 11 34 51

in my console.

+1 to enabling in Horizon to start (and not production), because we can have enough traffic there from external beta testers to get an idea of where changes can have a big impact.

Note on the commit message: it's probably OK to trim it down to just "enable this feature" with a note about it being only for Horizon. I don't think the rest is as useful for future reference. :)
